### ISSUE TITLE
Allow for future deprecation

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#30785]
We've long wanted the ability to say "deprecated since" for a *future* version of Rust. The motivation is that, currently, deprecation goes into effect immediately, even though in some cases the replacement for the deprecated item hasn't hit the stable channel yet. By "timing" the deprecation to a given release, we can make the decision and land the replacement, and have it automatically go into effect at the right time.

The idea would be that the deprecation does not warn until the compiler version is at least the version labeling the deprecation.

We could also pipe this information into rustdoc, to give a heads up on the forthcoming deprecation.

**Comment by aturon** [comment]
cc #30459

**Comment by steveklabnik** [comment]
Isn't this more of an RFC issue?

**Comment by aturon** [comment]
@steveklabnik I don't think so -- we don't need an RFC to do this, we just need to implement it.

**Comment by llogiq** [comment]
RFC [#1270](https://github.com/rust-lang/rfcs/blob/master/text/1270-deprecation.md) explicitly notes the possibility of future deprecation, but cautions that this implies a semver dependency. OTOH, `std` only has to care about rustc's version.

**Comment by steveklabnik** [comment]
Triage: no change.

**Comment by RalfJung** [comment]
I think it wold sometimes be useful to get warnings for using items deprecated in the future, see https://github.com/rust-lang/rust/issues/55892.
